pkg/upload/xcode-archive.go

```diff
@@ -0,0 +1,173 @@
+			_ = os.RemoveAll(tempDir)
+		}
+	}()
+
+	// Search for an .xcarchive in the specified paths
+	for _, path := range xcarchiveOptions.Path {
+		if filepath.Ext(path) == ".xcarchive" {
+			// If the path is directly an .xcarchive file, use it
+			xcarchivePath = path
+		} else if utils.IsDir(path) {
+			// If the path is a directory, explore it for an .xcarchive or an Xcode project/workspace
+			logger.Info(fmt.Sprintf("Searching for Xcode Archive in %s", path))
+
+			// Check if the directory contains an Xcode project or workspace
+			if ios.IsPathAnXcodeProjectOrWorkspace(path) {
+				// Set the project root based on Xcode project settings
+				xcarchiveOptions.Shared.ProjectRoot = ios.GetDefaultProjectRoot(path, xcarchiveOptions.Shared.ProjectRoot)
+				logger.Info(fmt.Sprintf("Setting `--project-root` from Xcode project settings: %s", xcarchiveOptions.Shared.ProjectRoot))
+
+				// Determine the default scheme for the project if not already provided
+				if xcarchiveOptions.Shared.Scheme == "" {
+					xcarchiveOptions.Shared.Scheme, err = ios.GetDefaultScheme(path)
+					if err != nil {
+						return err
+					}
+				}
+
+				// Retrieve Xcode build settings if a scheme is defined
+				if xcarchiveOptions.Shared.Scheme != "" {
+					buildSettings, err = ios.GetXcodeBuildSettings(path, xcarchiveOptions.Shared.Scheme, options.Upload.XcodeBuild.Configuration)
+					if err != nil {
+						logger.Warn(err.Error())
+					}
+				}
+
+				// Attempt to locate the latest .xcarchive associated with the project
+				xcarchiveLocation, err := ios.GetXcodeArchiveLocation()
+				if err != nil {
+					logger.Warn(fmt.Sprintf("Failed to get Xcode archive location: %s", err))
+					return err
+				}
+				xcarchivePath, err = ios.GetLatestXcodeArchive(xcarchiveLocation, xcarchiveOptions.Shared.Scheme)
+				if err != nil {
+					return err
+				}
+
+			} else {
+				// If not a project/workspace, search the directory for an .xcarchive
+				xcarchivePath, err = utils.FindFolderWithSuffix(path, ".xcarchive")
+				if err != nil {
+					logger.Warn(fmt.Sprintf("No xcarchive found in %s", path))
+					return err
+				}
+			}
+		}
+	}
+
+	// If no .xcarchive was found, return an error
+	if xcarchivePath == "" {
+		return fmt.Errorf("No xcarchive found in specified paths")
+	}
+	logger.Info(fmt.Sprintf("Found xcarchive at %s", xcarchivePath))
+
+	// Locate and process dSYM files within the .xcarchive
+	dwarfInfo, tempDir, err = ios.FindDsymsInPath(
+		xcarchivePath,
+		xcarchiveOptions.Shared.IgnoreEmptyDsym,
+		xcarchiveOptions.Shared.IgnoreMissingDwarf,
+		logger,
+	)
+	tempDirs = append(tempDirs, tempDir)
+	if err != nil {
+		return err
+	}
+	if len(dwarfInfo) == 0 {
+		return fmt.Errorf("No dSYM files found in: %s", xcarchivePath)
+	}
+	logger.Info(fmt.Sprintf("Found %d dSYM files in %s", len(dwarfInfo), xcarchivePath))
+
+	// Locate Info.plist if not explicitly provided
+	if plistPath == "" && options.ApiKey == "" && buildSettings != nil {
```

Comment:
IMO we shouldn't use the project settings (buildSettings) for anything other than locating the archive file as there's a reasonable expectation that we're doing the upload completely from the archive contents.

Am I right in thinking that `Info.plist` file is at the top of an archive? In which case, we should read this unless the plist path is specified in the command.